tests/zfs-tests/include/libtest.shlib

```diff
@@ -3662,39 +3663,200 @@ function ls_xattr # path
 	esac
 }
 
-function kstat # stat flags?
+#
+# Gather data from ZFS kstats.
+#
+#   kstat [-g] <stat> [<pool>]
+#
+# Returns the value of the <stat> kstat. If <pool> is provided, returns the
+# value of the pool-specific kstat.
+#
+#   $ kstat dbgmsg
+#   timestamp   message
+#   ...
+#
+#   $ kstat state garden
+#   ONLINE
+#
+# To get a single stat within a group or collection, separate the name with
+# '.' characters.
+#
+#   $ kstat dbufstats.cache_target_bytes
+#   131162693
+#
+#   $ kstat iostats.arc_read_bytes crayon
+#   184371908096
+#
+# -g is "group" mode. If the kstat is a group or collection, all stats in that
+# group are returned, one stat per line, key and value separated by a space.
+#
+#   $ kstat -g dbufstats
+#   metadata_cache_overflow 839250601
+#   metadata_cache_size_bytes_max 1617221632
+#   metadata_cache_size_bytes 1617074176
+#   ...
+#
+function _kstat_freebsd
 {
 	typeset stat=$1
-	typeset flags=${2-"-n"}
+	typeset pool=$2
+	typeset -i group_mode=$3
 
-	case "$UNAME" in
-	FreeBSD)
-		sysctl $flags kstat.zfs.misc.$stat
-		;;
-	Linux)
-		cat "/proc/spl/kstat/zfs/$stat" 2>/dev/null
-		;;
-	*)
-		false
-		;;
+	typeset oid=""
+	if [[ -z "$pool" ]] ; then
+		oid="kstat.zfs.misc.$stat"
+	else
+		oid="kstat.zfs.$pool.misc.$stat"
+	fi
+
+	# Calling sysctl on a "group" node will return everything under that
+	# node, so we have to inspect the first line to make sure we are
+	# getting back what we expect. For a single value, the key will have
+	# the name we requested, while for a group, the key will not have the
+	# name (group nodes are "opaque", not returned by sysctl by default.
+
+	if [[ $group_mode == 0 ]] ; then
+		sysctl -e "$oid" | awk -v oid="$oid" -v oidre="^$oid=" '
+			NR == 1 && $0 !~ oidre { exit 1 }
+			NR == 1 { print substr($0, length(oid)+2) ; next }
+			{ print }
```

Comment:
Did this `print` left from some debugging or it is needed for something?  Without group mode I suppose we should not even look on any following lines (which should not even be there) at all.

Reply:
Ah, OK.